src/python/pants/engine/intrinsics.py

```diff
@@ -120,13 +123,57 @@ async def run_id() -> RunId:
     return await native_engine.run_id()
 
 
+__SQUELCH_WARNING = "__squelch_warning"
+
+
+# NB: Call one of the helpers below, instead of calling this rule directly,
+#  to ensure correct application of restartable logic.
 @_uncacheable_rule
-async def interactive_process(
+async def _interactive_process(
     process: InteractiveProcess, process_execution_environment: ProcessExecutionEnvironment
 ) -> InteractiveProcessResult:
+    # This is a crafty way for a caller to signal into this function without a dedicated arg
+    # (which would confound the solver).  Note that we go via __dict__ instead of using
+    # setattr/delattr, because those error for frozen dataclasses.
+    if __SQUELCH_WARNING in process.__dict__:
+        del process.__dict__[__SQUELCH_WARNING]
+    else:
+        logging.warning(
+            "A plugin is calling `await Effect(InteractiveProcessResult, InteractiveProcess, "
+            "process)` directly. This will cause restarting logic not to be applied. "
+            "Use `await run_interactive_process(process)` or `await "
+            "run_interactive_process_in_environment(process, environment_name)` instead. "
+            "See src/python/pants/engine/intrinsics.py for more details."
+        )
     return await native_engine.interactive_process(process, process_execution_environment)
 
 
+async def run_interactive_process(process: InteractiveProcess) -> InteractiveProcessResult:
```

Comment:
Since we're now expecting people to call these directly, is `pants.engine.intrinsics.run_interactive_process` the right path? Maybe these should be in `pants.engine.process` instead?

Reply:
That would cause an import cycle between `pants.engine.intrinsics` and `pants.engine.process`. But in any case, since we want to tell you not to call the intrinsic directly, I figured it also made sense to have the thing you should call be adjacent...